indico/modules/rb_new/client/js/common/bookings/OccurrenceActionsDropdown.jsx

```diff
@@ -0,0 +1,212 @@
+/* This file is part of Indico.
+ * Copyright (C) 2002 - 2018 European Organization for Nuclear Research (CERN).
+ *
+ * Indico is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation; either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * Indico is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Indico; if not, see <http://www.gnu.org/licenses/>.
+ */
+
+import React from 'react';
+import PropTypes from 'prop-types';
+import {bindActionCreators} from 'redux';
+import {connect} from 'react-redux';
+import {Form as FinalForm, Field} from 'react-final-form';
+import {Button, Icon, Confirm, Dropdown, Form, Portal, TextArea} from 'semantic-ui-react';
+
+import {serializeDate} from 'indico/utils/date';
+import {ReduxFormField, formatters, validators as v} from 'indico/react/forms';
+import {Param, Translate} from 'indico/react/i18n';
+import * as bookingsActions from './actions';
+
+import './OccurrenceActionsDropdown.module.scss';
+
+
+class OccurrenceActionsDropdown extends React.Component {
+    static propTypes = {
+        booking: PropTypes.object.isRequired,
+        date: PropTypes.object.isRequired,
+        actions: PropTypes.exact({
+            changeBookingOccurrenceState: PropTypes.func.isRequired,
+            fetchBookingDetails: PropTypes.func.isRequired
+        }).isRequired,
+    };
+
+    constructor(props) {
+        super(props);
+        this.dropdownIconRef = React.createRef();
+    }
+
+    state = {
+        actionInProgress: false,
+        activeConfirmation: null,
+        dropdownOpen: false,
+        top: 0,
+        left: 0,
+    };
+
+
+    hideConfirm = () => {
+        this.setState({activeConfirmation: null});
+    };
+
+    showConfirm = (type) => {
+        this.setState({activeConfirmation: type});
+    };
+
+    changeOccurrenceState = async (action, data = {}) => {
+        const {
+            date,
+            booking: {id},
+            actions: {changeBookingOccurrenceState, fetchBookingDetails}
+        } = this.props;
+        const serializedDate = serializeDate(date);
+        this.setState({actionInProgress: true});
+        const rv = await changeBookingOccurrenceState(id, serializedDate, action, data).then(() => {
```

Comment:
Isn't this more correct?

        const rv = await changeBookingOccurrenceState(id, serializedDate, action, data))
        await fetchBookingDetails(id); // only await this one if you actually need to wait for it to complete
        this.setState({actionInProgress: false});  // this should probably happen even if there was an error
        return rv;

Reply:
actually, not sure if we even need to return anything here - all final-form cares about is that it's a promise (it uses that to track whether the form is currently submitting), but its return value only matters in case of an error, and since we don't use submitFailed...

Reply:
my very first thought when suggestions async/await was actually this:

```js
await changeBookingOccurrenceState(id, serializedDate, action, data))
await fetchBookingDetails(id);
this.setState({actionInProgress: false});
```

Reply:
I just saw it now, I can update a bit later :)